pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -165,46 +190,305 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
 	return predicateMetadata
 }
 
+func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*schedulernodeinfo.NodeInfo) *topologyPairsPodSpreadMap {
+	// we have feature gating in APIserver to strip the spec
+	// so don't need to re-check feature gate, just check length of constraints
+	constraints := getHardTopologySpreadConstraints(pod)
+	if len(constraints) == 0 {
+		return nil
+	}
+
+	allNodeNames := make([]string, 0, len(nodeInfoMap))
+	for name := range nodeInfoMap {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	var lock sync.Mutex
+	var firstError error
+
+	topologyPairsPodSpreadMap := newTopologyPairsPodSpreadMap()
+
+	appendTopologyPairsMaps := func(toAppend *topologyPairsMaps) {
+		lock.Lock()
+		topologyPairsPodSpreadMap.appendMaps(toAppend)
+		lock.Unlock()
+	}
+	catchError := func(err error) {
+		lock.Lock()
+		if firstError == nil {
+			firstError = err
+		}
+		lock.Unlock()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	processNode := func(i int) {
+		nodeInfo := nodeInfoMap[allNodeNames[i]]
+		node := nodeInfo.Node()
+		if node == nil {
+			catchError(fmt.Errorf("node not found"))
```

Comment:
should node name be included in the message ?

Reply:
Yes. I will update in the original PR which first introduced this change, and rebase here.